### PR TITLE
[ADD] project: portal/sharelink use backend views

### PR DESCRIPTION
**Pull request by Zynton**
https://www.odoo.com/web#id=1866286&model=project.task&view_type=form&menu_id=5195

Purpose
=======
Allow external users to receive an access link to a give project, and to display the kanban view on the frontend, with limited access

### CODE REVIEW

File: addons/portal/models/portal_mixin.py [hunk 1]
```diff
@@ -32,9 +33,15 @@ def _portal_ensure_token(self):
...
+    def _portal_ensure_token_edit(self):
+        """ Get the current record edit & readonly access tokens """
+        self._portal_ensure_token()
+        self.edit_token = self.edit_token if self.edit_token else str(uuid.uuid4())
+        return self.edit_token
+
+    def _get_share_url(self, redirect=False, signup_partner=False, pid=None, portaledit=False):
         """
-        Build the url of the record  that will be sent by mail and adds additional parameters such as
+        Build the url of the record that will be sent by mail and adds additional parameters such as
```
Comment (dbeguin) [inline addons/portal/models/portal_mixin.py]: unnecessary diff

File: addons/project/controllers/dataset.py [hunk 2]
```diff
@@ -0,0 +1,351 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import api, http
+from odoo.addons.web.controllers import main
+from odoo.http import request
+from odoo.tools import consteq
+
```
Comment (dbeguin) [inline addons/project/controllers/dataset.py]: linting : 2 empty lines before class declaration

File: addons/project/controllers/mail.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import http, SUPERUSER_ID
+from odoo.http import request
+from odoo.addons.mail.controllers.main import MailController
+
```
Comment (dbeguin) [inline addons/project/controllers/mail.py]: linting : 2 empty lines before class declaration

File: addons/project/controllers/portal.py [hunk 4]
```diff
@@ -31,15 +31,15 @@ def _prepare_portal_layout_values(self):
     def _project_get_page_view_values(self, project, access_token, **kwargs):
         values = {
             'page_name': 'project',
-            'project': project,
+            'project_id': project.id,
```
Comment (dbeguin) [inline addons/project/controllers/portal.py]: For future development, it could be good to keep the object and not only the id. because if you need, later, in the xml template to add some values from the project, you will have to revert your changes anyway. In the template 'portal_my_project', you can use project.id instead of project_id

File: addons/project/controllers/portal.py [hunk 5]
```diff
@@ -97,7 +97,8 @@ def portal_my_project(self, project_id=None, access_token=None, **kw):
     def _task_get_page_view_values(self, task, access_token, **kwargs):
         values = {
             'page_name': 'task',
-            'task': task,
+            'project_id': task.project_id.id,
+            'task_id': task.id,
```
Comment (dbeguin) [inline addons/project/controllers/portal.py]: Same remark than project_id -> use object instead of id

File: addons/project/controllers/portal.py [hunk 6]
```diff
@@ -206,7 +207,56 @@ def portal_my_task(self, task_id, access_token=None, **kw):
...
+            try:
+                # Task can be accessed with its project's access token as well
+                task_sudo = request.env['project.task'].sudo().browse(task_id)
+                self._document_check_access('project.project', task_sudo.project_id.id, access_token)
+            except (AccessError, MissingError):
+                return request.redirect('/my')
 
         values = self._task_get_page_view_values(task_sudo, access_token, **kw)
         return request.render("project.portal_my_task", values)
+
```
Comment (dbeguin) [inline addons/project/controllers/portal.py]: lint : unecessary line

File: addons/project/static/src/css/project_portal.css [hunk 7]
```diff
@@ -0,0 +1,21 @@
+.o_project_portal_iframe {
+    background-color: #fff;
+    border: 0;
+    overflow-y: hidden;
+    width: 100%;
+}
+.o_project_portal_iframe_content {
```
Comment (dbeguin) [inline addons/project/static/src/css/project_portal.css]: use multiple class for same properties notation
.o_project_portal_iframe {
    border: 0;
    width: 100%;
}
o_project_portal_iframe, .o_project_portal_iframe_content {
    background-color: #fff;
    overflow-y: hidden;
}

File: addons/project/static/src/js/portal_project.js [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+            var portalForm = new PortalForm(params);
+            portalForm.appendTo($target);
+            break;
+        default:
+            return $.Deferred().reject("No view type was specified in the params passed to 'portal-project'");
+    };
+
+    setInterval(function() {
+        resizeIFrame($(document).height())
+    }, 100);
```
Comment (dbeguin) [inline addons/project/static/src/js/portal_project.js]: In general we try to avoid using intervals or timeout in general.
also, is it really necessary to resize the iframe every 0.1 sec ?
there are other ways to wait for something to be ready before executing other instructions (callback, promise or.. see with js team, they are better than me on this :D)

File: addons/project/views/project_portal_templates.xml [hunk 9]
```diff
@@ -162,84 +127,31 @@
...
-                        <t t-set="hash" t-value="hash"/>
-                    </t>
-                </div>
-            </div>
-        </t>
+                <t t-call-assets="web.assets_common"/>
+                <t t-call-assets="web.assets_backend"/>
+                <t t-call-assets="web_editor.summernote"/>
+                <t t-call-assets="web_editor.assets_editor"/>
+
```
Comment (dbeguin) [inline addons/project/views/project_portal_templates.xml]: unecessary empty line

File: addons/project/views/project_portal_views.xml [hunk 10]
```diff
@@ -0,0 +1,81 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+    <record id="portal_view_task_kanban" model="ir.ui.view">
```
Comment (dbeguin) [inline addons/project/views/project_portal_views.xml]: guidelines : id would be "portal_task_view_kanban" (module_model_view/action_viewType)
You can also adapt all the other record id of the file

File: addons/web/controllers/main.py [hunk 11]
```diff
@@ -948,11 +948,12 @@ def call_button(self, model, method, args, domain_id=None, context_id=None):
 
     @http.route('/web/dataset/resequence', type='json', auth="user")
     def resequence(self, model, ids, field='sequence', offset=0):
-        """ Re-sequences a number of records in the model, by their ids
+        """ Re-sequences a number of records in a model, by their ids
```
Comment (dbeguin) [inline addons/web/controllers/main.py]: unnecessary diff

File: addons/web/controllers/main.py [hunk 12]
```diff
@@ -961,7 +962,7 @@ def resequence(self, model, ids, field='sequence', offset=0):
                            starting the resequencing from an arbitrary number,
                            defaults to ``0``
         """
-        m = request.env[model]
+        m = request.env[model] if isinstance(model, str) else model
```
Comment (dbeguin) [inline addons/web/controllers/main.py]: why adding a model parameter as we can retrieve it already using request.env[model] ?

File: addons/portal/models/portal_mixin.py [hunk 13]
```diff
@@ -13,7 +13,8 @@ class PortalMixin(models.AbstractModel):
     access_url = fields.Char(
         'Portal Access URL', compute='_compute_access_url',
         help='Customer Portal URL')
-    access_token = fields.Char('Security Token', copy=False)
+    access_token = fields.Char('Security Token (readonly)', copy=False, default=str(uuid.uuid4()))
```
[the remark this answers is not in the document]
Comment (Zinston) [reply addons/portal/models/portal_mixin.py]: The way it is right now is that you always have a public link in the project form, with a readonly token. Hence the token needs always be generated. We're not the only ones to have our doubts over that behavior so the discussion's open and specs are likely to change soon. In the meantime I believe I'm going to have to leave this the way it is.

File: addons/project/static/src/js/portal_project.js (hunk 8 above)
Comment (Zinston) [reply addons/project/static/src/js/portal_project.js]: I'll see if anyone at the JS team has a better suggestion. I also didn't like writing this ;-)
Also, 0.1s delay is maybe a bit excessive indeed, I just wanted to make it seamless :-)

File: addons/web/controllers/main.py (hunk 12 above)
Comment (Zinston) [reply addons/web/controllers/main.py]: I'm not sure I understand your question.
What happens here is resequence assumes the model parameter to be a string but for the purposes of my overrides in dataset.py (_sudo_resequence) I'd like it to be able to be the model itself. That is so I can pass the model as sudo, which I can't do if the model is requested in this way.
It's the least intrusive way I found to do this.

**Comment by Zinston** [comment]
Thanks @dbeguin for your review!
Concerning the iFrames, it was decided in discussion with GED indeed. It has the advantage of completely insolating the backend views that I'm inserting from the rest of the portal, which avoids us a bunch of JS and CSS conflicts. It's possible to do it without the iFrame but that would require a very large rewrite of the common/back/frontend assets, as well as a bridge module for Website. Also, iFrames are used in other places in Odoo - for reporting for instance - and don't seem to cause many issues.
I first worked without iFrames, believing them to be evil as well... But then GED was pretty convincing in shrugging that off :-)

File: addons/project/static/src/css/project_portal.css [hunk 14]
```diff
@@ -0,0 +1,19 @@
+.o_project_portal_iframe {
+    border: 0;
+    width: 100%;
+}
+o_project_portal_iframe, .o_project_portal_iframe_content {
```
Comment (dbeguin) [inline addons/project/static/src/css/project_portal.css]: Missing "." just before o_project_portal_iframe

File: addons/project/static/src/js/portal_form.js [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+            },
+
+            /**
+             * Prevent the onclick event from bubbling up
+             * if the field links to project.project
+             *
+             * @override
+             * @private
+             */
+            _onClick: function(event) {
```
Comment (adr-odoo) [inline addons/project/static/src/js/portal_form.js]: event? docstring

File: addons/portal/models/portal_mixin.py [hunk 16]
```diff
@@ -145,3 +154,19 @@ def _sign_token(self, pid):
...
+        """
+        self.ensure_one()
+        portal_access = portal_edit = False
+        if token:
+            record_sudo = self.sudo()
+            if record_sudo.access_token and consteq(record_sudo.access_token, token):
+                portal_access = True
+            if record_sudo.edit_token and consteq(record_sudo.edit_token, token):
+                portal_access = portal_edit = True
+        return portal_access, portal_edit
```
Comment (dbeguin) [inline addons/portal/models/portal_mixin.py]: You could avoid to test portal_access by testing portal_edit first. If portal edit = True, no need to test portal_access.
also, you could use direct assignation

        self.ensure_one()
        record_sudo = self.sudo()
        portal_access = portal_edit = token and record_sudo.edit_token and consteq(record_sudo.edit_token, token)
        if not portal_access:
            portal_access = token and record_sudo.access_token and consteq(record_sudo.access_token, token)
        return portal_access, portal_edit

File: addons/project/static/src/js/portal_project.js (hunk 8 above)
Comment (Zinston) [reply addons/project/static/src/js/portal_project.js]: Well that did create a problem with the mobile version but I fixed that as well. `document.body` didn't take the kanban cards' heights into account as their position was absolute. My (hacky, I admit) fix has been to set their position to relative if active and absolute if not.

```javascript
function adjustPositions() {
    $('.o_kanban_group').each(function () {
        if ($('.o_current').attr('data-id') === $(this).attr('data-id')) {
            $(this).css('position', 'relative');
        } else {
            $(this).css('position', 'absolute');
        };
    });
};
```
and
```css
@media (max-width: 767.98px) {
    .o_kanban_view.o_kanban_grouped[class] .o_kanban_group:not(.o_column_folded) {
        top: 0;
    }
}
```

File: addons/mail/models/mail_thread.py [hunk 17]
```diff
@@ -620,7 +620,7 @@ def _message_track(self, tracked_fields, initial):
         return changes, tracking_value_ids
 
     @api.multi
-    def message_track(self, tracked_fields, initial_values):
+    def message_track(self, tracked_fields, initial_values, **kwargs):
```
Comment (Zinston) [inline addons/mail/models/mail_thread.py]: verify if possible to only accept author_id, not all kwargs to avoid exposing other fields

File: addons/portal/models/portal_mixin.py [hunk 18]
```diff
@@ -14,7 +15,8 @@ class PortalMixin(models.AbstractModel):
     access_url = fields.Char(
         'Portal Access URL', compute='_compute_access_url',
         help='Customer Portal URL')
-    access_token = fields.Char('Security Token', copy=False)
+    access_token = fields.Char('Security Token (readonly)', copy=False)
```
Comment (Zinston) [inline addons/portal/models/portal_mixin.py]: try to only use one token, which will simplify the logic in many places: no need for dual _check_token, etc.

File: addons/project/controllers/public_project.py [hunk 19]
```diff
@@ -0,0 +1,369 @@
...
+        """Check permissions before calling a given method.
+
+        :param method_to_call: (str) The name of the method to call if access is granted.
+        :param submethod: (str) The name of the method that the method to call will call
+                                if access is granted (if method_to_call == call or call_kw).
+                                Defaults to None.
+        :param kw: (dict) All args and kwargs passed to the method to call.
+        :return: (any) The result of the method to call or None if no access was granted.
+        """
+        read_words = ['read', 'load', 'default', 'get', 'search',
```
Comment (Zinston) [inline addons/project/controllers/public_project.py]: Extract the actual list of methods (not substring/keywords) and move it to dedicated methods that could be overridden in other modules if needed. This will make sure we never allow unsafe methods to be called, or private methods, etc.

File: addons/project/controllers/public_project.py [hunk 20]
```diff
@@ -0,0 +1,369 @@
...
+                     'message', 'mail',  # we want to allow readonly users to use the chatter
+                     'onchange', 'export_data']
```
Comment (Zinston) [inline addons/project/controllers/public_project.py]: remove export_data, set, get, ... if not needed

File: addons/project/controllers/public_project.py [hunk 21]
```diff
@@ -0,0 +1,369 @@
...
+        edit_words = ['write', 'resequence', 'set']
+
+        if any(word in (submethod or method_to_call) for word in read_words):
```
Comment (Zinston) [inline addons/project/controllers/public_project.py]: Always check method_to_call and submethod separately. Only allow submethod if method_to_call is call/call_kw

File: addons/project/models/project.py [hunk 22]
```diff
@@ -173,6 +179,7 @@ def _get_default_favorite_user_ids(self):
         return [(6, 0, [self.env.uid])]
 
     name = fields.Char("Name", index=True, required=True, track_visibility='onchange')
+    access_token = fields.Char('Security Token (readonly)', copy=False, default=str(uuid.uuid4()))
```
[the remark this answers is not in the document]
Comment (Zinston) [reply addons/project/models/project.py]: (Above comment was by ODO)
Fixed the static token issue. As for generating it only when the project is shared, waiting for green light from JCR (spec change).

File: addons/project/views/project_portal_views.xml [hunk 23]
```diff
@@ -0,0 +1,100 @@
...
+    </record>
+
+    <record id="portal_view_task_search_form" model="ir.ui.view">
+        <field name="name">portal.project.task.search.form</field>
+        <field name="model">project.task</field>
+        <field name="arch" type="xml">
+            <search>
+                <field name="name" string="Tasks"/>
+                <field name="tag_ids"/>
+                <field name="user_id"/>
```
[the remark this answers is not in the document]
Comment (Zinston) [reply addons/project/views/project_portal_views.xml]: (Above comment was by ODO)
Do we mean to filter for the user's own tasks then and not allow a search by user name?

[the remark this answers is not in the document]
Comment (Zinston) [reply addons/project/views/project_portal_views.xml]: I did the `filter_domain` for the moment.

File: addons/project/controllers/public_project.py [hunk 24]
```diff
@@ -0,0 +1,369 @@
...
+                                                submethod=submethod,
+                                                access_info=access_info,
+                                                **kw)
+        if any(word in (submethod or method_to_call) for word in edit_words):
+            return self._check_and_perform_write(method_to_call,
+                                                 submethod=submethod,
+                                                 access_info=access_info,
+                                                 **kw)
+        if 'create' in (submethod or method_to_call):
+            return self._check_and_perform_create(method_to_call,
```
[the remark this answers is not in the document]
Comment (Zinston) [reply addons/project/controllers/public_project.py]: (Above comment was by ODO)
Not authorizing operations on the 'res.user' model implies removing the 'assign to' field in task's quick create form. 'assigned to' having a default value, this has the effect of assigning the task to an arbitrary user. What behavior do we wish here?

[the remark this answers is not in the document]
Comment (Zinston) [reply addons/project/controllers/public_project.py]: Done. Regarding the domains, I did roughly this:
```python
for d in domain:
    if not isinstance(d, str):
        is_field_ok = self._is_whitelisted(d[0])
        has_subexpr = isinstance(d[0], str) and '.' in d[0]
        if not is_field_ok or has_subexpr:
            raise AccessError("nope")
```
`has_subexpr` is not necessary as no field in the whitelist has a '.' sub-expression, but I included it anyway in case someone ever adds one of those to the whitelist. In any case the fact that I'm checking all d[0]'s against a whitelist of fields makes that I don't need to whitelist domains in particular.

**Comment by tde-banana-odoo** [comment]
Project sharing ha been done since then, closing.
